### PR TITLE
Improved API for loading Terrier indices into memory

### PR DESCRIPTION
**Pull request by cmacdonald**
(WIP)

### CODE REVIEW

**Comment by seanmacavaney** [comment]
What's going on here -- can I have a bit of context?

**Comment by cmacdonald** [comment]
Woops. The pr omitted my explanations or any meaningful title. See https://gist.github.com/cmacdonald/c12ddacd73ad379b3b0f7a8b7cf1d080

File: pyterrier/bootstrap.py [hunk 1]
```diff
@@ -5,6 +5,87 @@
...
+                continue
+            clz = pindex.getIndexProperty(f"index.{s}.class", "notfound")
+            if not clz in REWRITES[s]:
+                raise ValueError(f"Cannot load structure {s} into memory, underlying class {clz} is not supported")
+
+            # we only reload an index structure if a property has changed
+            dirty = False
+            for k, v in REWRITES[s][clz].items():
+                if pindex.getIndexProperty(k, "notset") != v:
+                    print("Rewriting " + k)
```
Comment (cmacdonald) [inline pyterrier/bootstrap.py:56]: remove the print.

File: pyterrier/bootstrap.py [hunk 2]
```diff
@@ -1,10 +1,120 @@
...
 SAVED_FNS=[]
 
+class IndexFactory:
+    """
+    This class "shades" the native Terrier `IndexFactory <http://terrier.org/docs/current/javadoc/org/terrier/structures/IndexFactory.html>`_ class - it offers essential the same API,
+    except that the ``of()`` method contains a memory kwarg, that can be used to load additional index data structures into memory. 
+    """
+
+    @staticmethod
+    def _load_into_memory(index, structures=['lexicon', 'direct', 'inverted', 'meta'], load=False):
```
Comment (tonellotto) [inline pyterrier/bootstrap.py:16]: What about document index? It is accessed for getting doc lens, so quite often during ranking.

Comment (cmacdonald) [reply pyterrier/bootstrap.py:16]: document lengths are always in memory. the choice of memory or not relates to the other information, specifically the bits used to access the direct index

**Pull request merged**